libvast/include/vast/detail/collect.hpp

```diff
@@ -0,0 +1,27 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/detail/generator.hpp>
+
+namespace vast::detail {
+
+/// A utility function to collect all results produced by a
+/// `vast::detail::generator<T>` into a suitable standard container.
+/// Example:
+///     auto g = vast::detail::generator<int>{};
+///     auto v = vast::detail::collect<std::vector>(g);
+template <template <class, class> class Container, class Generated>
+Container<Generated, std::allocator<Generated>>
+collect(vast::detail::generator<Generated> g) {
+  Container<Generated, std::allocator<Generated>> result;
+  for (auto& x : g)
+    result.push_back(x);
```

Comment:
The more portable way to implement this such that it works for other containers as well is to use a back inserter, e.g.,

```
std::move(g.begin(), g.end(), std::back_inserter(result));
```

Reply:
Is it safe to move out of `g.begin()`? It returns a reference, so my hunch was probably not.

Reply:
It is safe. This being a reference is an artifact of our coroutine implementation being based on an early draft of the proposal for C++23. There's also sc-32185 to make the behavior a bit more intuitive.

Reply:
I've appended an implementation of `sc-32185` in a separate commit, to make the ownership semantics explicit. (the story is technically only a discussion story, but I guess we can discuss as part of the review in this PR; I expect the general thrust to be non-controversial)

Reply:
Thank you so much! I think we can archive that story then?

Also wrt sentinel support: I think that's just missing from the standard library implementation, the relaxed iterator specification was in C++20 already iirc.

Marking as resolved.